### PR TITLE
gitee lifecycle plugin 

### PR DESCRIPTION
**Pull request by xwz123**
1. support close pr and issue command.
2. support reopen issue command.

### CODE REVIEW

File: prow/gitee/client.go [hunk 1]
```diff
@@ -440,6 +448,35 @@ func (c *client) ReplacePRAllLabels(owner, repo string, number int, labels []str
...
+		State: "closed",
+	}
+	_, err := c.UpdateIssue(owner, number, opt)
+	return formatErr(err, "close issue")
+}
+
+func (c *client) ReopenIssue(owner, repo string, number string) error {
+	opt := sdk.IssueUpdateParam{
+		Repo:  repo,
+		State: "open",
```
Comment (zengchen1024) [inline prow/gitee/client.go:464]: Is there any variable from gitee library to be on behalf of `open`

File: prow/gitee/client.go [hunk 2]
```diff
@@ -440,6 +448,35 @@ func (c *client) ReplacePRAllLabels(owner, repo string, number int, labels []str
 	return formatErr(err, "replace pr labels")
 }
 
+
+func (c *client) CloseIssue(owner, repo string, number string) error {
+	opt := sdk.IssueUpdateParam{
+		Repo:  repo,
+		State: "closed",
```
Comment (zengchen1024) [inline prow/gitee/client.go:455]: Is there any variable from gitee library to be on behalf of `close`

File: prow/gitee/client.go [hunk 3]
```diff
@@ -297,6 +297,14 @@ func (c *client) RemovePRLabel(org, repo string, number int, label string) error
 	return formatErr(err, "remove label of pr")
 }
 
+func (c *client) ClosePR(org, repo string, number int) error {
+	opt := sdk.PullRequestUpdateParam{
+		State:"closed",
```
Comment (zengchen1024) [inline prow/gitee/client.go:302]: Is there any variable from gitee library to be on behalf of `closed`

File: prow/gitee-plugins/lifecycle/reopen.go [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+
+func handleReopen(gc reopenClient, log *logrus.Entry, e *sdk.NoteEvent) error {
+	if isPr(*e.NoteableType) || e.Issue.State != "closed" {
+		return nil
+	}
+	if !reopenRe.MatchString(e.Comment.Body) {
+		return nil
+	}
+	org, repo, err := plugins.GetOwnerAndRepoByEvent(e)
+	if err != nil {
```
Comment (zengchen1024) [inline prow/gitee-plugins/lifecycle/reopen.go:29]: ignore error

File: prow/gitee-plugins/lifecycle/reopen.go [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+	isAuthor := e.Issue.User.Login == commentAuthor
+	isCollaborator, err := gc.IsCollaborator(org, repo, commentAuthor)
+	if err != nil {
+		log.WithError(err).Errorf("Failed IsCollaborator(%s, %s, %s)", org, repo, commentAuthor)
+	}
+
+	number := e.Issue.Number
+	// Only authors and collaborators are allowed to reopen issues or PRs.
+	if !isAuthor && !isCollaborator {
+		response := "You can't reopen an issue/PR unless you authored it or you are a collaborator."
```
Comment (zengchen1024) [inline prow/gitee-plugins/lifecycle/reopen.go:42]: You can't reopen an issue or a PR unless you are the author of it or a collaborator

File: prow/gitee-plugins/lifecycle/reopen.go [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+		log.Infof("Commenting \"%s\".", response)
```
Comment (zengchen1024) [inline prow/gitee-plugins/lifecycle/reopen.go:43]: don't write log

File: prow/gitee-plugins/lifecycle/reopen.go (hunk 5 above)
Comment (zengchen1024) [reply prow/gitee-plugins/lifecycle/reopen.go:42]: s/!isAuthor/e.Issue.User.Login != commentAuthor

File: prow/gitee-plugins/lifecycle/reopen.go [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+var reopenRe = regexp.MustCompile(`(?mi)^/reopen\s*$`)
+
+type reopenClient interface {
+	IsCollaborator(owner, repo, login string) (bool, error)
+	CreateGiteeIssueComment(org, repo string, number string, comment string) error
+	ReopenIssue(owner, repo string, number string) error
+}
+
+func handleReopen(gc reopenClient, log *logrus.Entry, e *sdk.NoteEvent) error {
+	if isPr(*e.NoteableType) || e.Issue.State != "closed" {
```
Comment (zengchen1024) [inline prow/gitee-plugins/lifecycle/reopen.go:22]: isPr doesn't mean the event is for issue.

File: prow/gitee-plugins/lifecycle/reopen.go [hunk 8]
```diff
@@ -0,0 +1,55 @@
...
+	// Only authors and collaborators are allowed to reopen issues or PRs.
+	if !isAuthor && !isCollaborator {
+		response := "You can't reopen an issue/PR unless you authored it or you are a collaborator."
+		log.Infof("Commenting \"%s\".", response)
+		return gc.CreateGiteeIssueComment(
+			org, repo, number, originp.FormatResponseRaw(e.Comment.Body, e.Comment.HtmlUrl, commentAuthor, response))
+	}
+	err = gc.ReopenIssue(org, repo, number)
+	if err != nil {
+		return err
```
Comment (zengchen1024) [inline prow/gitee-plugins/lifecycle/reopen.go:49]: if err = gc.ReopenIssue(org, repo, number); err != nil {

File: prow/gitee-plugins/lifecycle/reopen.go (hunk 5 above)
Comment (zengchen1024) [reply prow/gitee-plugins/lifecycle/reopen.go:42]: if !isAuthor {
	isCollaborator, err := gc.IsCollaborator(org, repo, commentAuthor)
	if err != nil {
		log.WithError(err).Errorf("Failed IsCollaborator(%s, %s, %s)", org, repo, commentAuthor)
	}
        if !isCollaborator{
}
}

File: prow/gitee-plugins/lifecycle/lifecycle.go [hunk 9]
```diff
@@ -0,0 +1,84 @@
...
+func (l *lifecycle) HelpProvider(_ []prowConfig.OrgRepo) (*pluginhelp.PluginHelp, error) {
+	pluginHelp := &pluginhelp.PluginHelp{
+		Description: "Close an issue or PR",
+	}
+	pluginHelp.AddCommand(pluginhelp.Command{
+		Usage:       "/close",
+		Featured:    false,
+		Description: "Closes an issue or PullRequest.",
+		Examples:    []string{"/close"},
+		WhoCanUse:   "Authors and collaborators on the repository can trigger this command.",
```
Comment (zengchen1024) [inline prow/gitee-plugins/lifecycle/lifecycle.go:36]: collaborators of the repository

File: prow/gitee-plugins/lifecycle/lifecycle.go [hunk 10]
```diff
@@ -0,0 +1,84 @@
...
+}
+
+func (l *lifecycle) handleNoteEvent(e *sdk.NoteEvent, log *logrus.Entry) error {
+	funcStart := time.Now()
+	defer func() {
+		log.WithField("duration", time.Since(funcStart).String()).Debug("Completed handleNoteEvent")
+	}()
+
+	eType := *(e.NoteableType)
+	if *(e.Action) != "comment" || (!isPr(eType) && eType != "Issue") {
```
Comment (zengchen1024) [inline prow/gitee-plugins/lifecycle/lifecycle.go:68]: !isPr(eType) && eType != "Issue"
can new a function of isIssue

**Review by zengchen1024 — CHANGES_REQUESTED** [review]
need change

File: prow/gitee-plugins/lifecycle/reopen.go (hunk 7 above)
Comment (xwz123) [reply prow/gitee-plugins/lifecycle/reopen.go:22]: Already judged at the event entrance

File: prow/gitee-plugins/lifecycle/close.go [hunk 11]
```diff
@@ -0,0 +1,77 @@
...
+}
+
+func closeIssue(gc closeClient, log *logrus.Entry, e *sdk.NoteEvent) error {
+	org, repo := giteep.GetOwnerAndRepoByEvent(e)
+	commentAuthor := e.Comment.User.Login
+	author := e.Issue.User.Login
+	number := e.Issue.Number
+
+	if !isAuthorOrCollaborator(org, repo, author, commentAuthor, gc.IsCollaborator, log) {
+		response := "You can't close an  issue unless you authored it or you are a collaborator."
```
Comment (zengchen1024) [inline prow/gitee-plugins/lifecycle/close.go:32]: You can't close an issue unless you are the author of it or a collaborator

File: prow/gitee-plugins/lifecycle/close.go [hunk 12]
```diff
@@ -0,0 +1,77 @@
...
+	if e.PullRequest.State != gitee.StatusOpen && closeRe.MatchString(e.Comment.Body) {
+		return nil
+	}
+	org, repo := giteep.GetOwnerAndRepoByEvent(e)
+	commentAuthor := e.Comment.User.Login
+	author := e.PullRequest.User.Login
+	number := int(e.PullRequest.Number)
+
+	if !isAuthorOrCollaborator(org, repo, author, commentAuthor, gc.IsCollaborator, log) {
+		response := "You can't reopen an issue unless you are the author of it or a collaborator"
```
Comment (zengchen1024) [inline prow/gitee-plugins/lifecycle/close.go:53]: ditto

File: prow/gitee/type.go [hunk 13]
```diff
@@ -0,0 +1,23 @@
...
+)
+
+//IsPullRequest Determine whether it is a PullRequest
+func IsPullRequest(noteType string) bool {
+	return noteType == "PullRequest"
+}
+
+//IsIssue Determine whether it is a issue
+func IsIssue(noteType string) bool {
+	return noteType == "issue"
```
Comment (zengchen1024) [inline prow/gitee/type.go:18]: Issue

**Review by zengchen1024 — COMMENTED** [review]
need change

File: prow/gitee-plugins/lifecycle/lifecycle.go [hunk 14]
```diff
@@ -0,0 +1,97 @@
...
+	}()
+
+	if !gitee.IsCreateCommentEvent(*(e.Action)) {
+		log.Debug("Event is not a creation of a comment for PR or issue, skipping.")
+		return nil
+	}
+
+	eType := *(e.NoteableType)
+	if gitee.IsPullRequest(eType) {
+		if err := closePullRequest(l.gec, log, e); err != nil {
```
Comment (zengchen1024) [inline prow/gitee-plugins/lifecycle/lifecycle.go:74]: return closePullRequest(l.gec, log, e)

File: prow/gitee/noteevent.go [hunk 15]
```diff
@@ -0,0 +1,78 @@
...
+	return *(ne.NoteableType) == "PullRequest"
+}
+
+//IsIssue Determine whether it is a issue
+func (ne *NoteEvent) IsIssue() bool {
+	return *(ne.NoteableType) == "Issue"
+}
+
+//IsCreateCommentEvent Determine whether an note event is create a comment
+func (ne *NoteEvent) IsCreateCommentEvent() bool {
```
Comment (zengchen1024) [inline prow/gitee/noteevent.go:25]: s/IsCreateCommentEvent/IsCreatingComment

File: prow/gitee/noteevent.go [hunk 16]
```diff
@@ -0,0 +1,78 @@
...
+	return *(ne.Action) == "comment"
+}
+
+//IssueIsClosed whether the status is close  of the issue
+func (ne *NoteEvent) IssueIsClosed() bool {
```
Comment (zengchen1024) [inline prow/gitee/noteevent.go:30]: s/IssueIsClosed/IsIssueClosed/

File: prow/gitee/noteevent.go [hunk 17]
```diff
@@ -0,0 +1,78 @@
...
+	return ne.Issue.State == StatusClosed
+}
+
+//IssueIsOpen whether the status is open  of the issue
+func (ne *NoteEvent) IssueIsOpen() bool {
```
Comment (zengchen1024) [inline prow/gitee/noteevent.go:35]: s/IssueIsOpen/IsIssueOpened

File: prow/gitee/noteevent.go [hunk 18]
```diff
@@ -0,0 +1,78 @@
...
+	return ne.Issue.State == StatusOpen
+}
+
+//IssueIsOpen whether the status is open  of the PullRequest
+func (ne *NoteEvent) PRIsOpen() bool {
```
Comment (zengchen1024) [inline prow/gitee/noteevent.go:40]: s/PRIsOpen/IsPROpened

File: prow/gitee/noteevent.go [hunk 19]
```diff
@@ -0,0 +1,78 @@
...
+	return ne.PullRequest.State == StatusOpen
+}
+
+//CommenterIsIssueAuthor Whether the author of the comment is the author of the issue
+func (ne *NoteEvent) CommenterIsIssueAuthor() bool {
```
Comment (zengchen1024) [inline prow/gitee/noteevent.go:45]: s/CommenterIsIssueAuthor/IsCommenterSameAsIssueAuthor

File: prow/gitee/noteevent.go [hunk 20]
```diff
@@ -0,0 +1,78 @@
...
+	return ne.Comment.User.Login == ne.Issue.User.Login
+}
+
+//CommenterIsPRAuthor Whether the author of the comment is the author of the PR
+func (ne *NoteEvent) CommenterIsPRAuthor() bool {
```
Comment (zengchen1024) [inline prow/gitee/noteevent.go:50]: s/CommenterIsPRAuthor/IsCommenterSameAsPRAuthor/

File: prow/gitee/noteevent.go [hunk 21]
```diff
@@ -0,0 +1,78 @@
...
+	return ne.Comment.User.Login == ne.PullRequest.User.Login
+}
+
+//CommenterIsAuthor Whether the author of the comment is the author of the PR or issue
+func (ne *NoteEvent) CommenterIsAuthor() bool {
```
Comment (zengchen1024) [inline prow/gitee/noteevent.go:55]: delete

File: prow/gitee/noteevent.go [hunk 22]
```diff
@@ -0,0 +1,78 @@
...
+		return ne.CommenterIsIssueAuthor()
+	}
+	if ne.IsPullRequest() {
+		return ne.CommenterIsPRAuthor()
+	}
+	return false
+}
+
+//GetCommenter Return to the author of the comment
+func (ne *NoteEvent) GetCommenter() string {
```
Comment (zengchen1024) [inline prow/gitee/noteevent.go:66]: s/GetCommenter/Commenter

File: prow/gitee/noteevent.go [hunk 23]
```diff
@@ -0,0 +1,78 @@
...
+	return ne.Comment.User.Login
+}
+
+//GetPRNumber Return to the number of the PR
+func (ne *NoteEvent) GetPRNumber() int {
```
Comment (zengchen1024) [inline prow/gitee/noteevent.go:71]: s/GetPRNumber/PRNumber

File: prow/gitee/noteevent.go [hunk 24]
```diff
@@ -0,0 +1,78 @@
...
+	return int(ne.PullRequest.Number)
+}
+
+//GetIssueNumber Return to the number of the issue
+func (ne *NoteEvent) GetIssueNumber() string {
```
Comment (zengchen1024) [inline prow/gitee/noteevent.go:76]: s/GetIssueNumber/IssueNumber/

File: prow/gitee-plugins/lifecycle/reopen.go [hunk 25]
```diff
@@ -0,0 +1,31 @@
...
+			org, repo, number, plugins.FormatResponseRaw(e.Comment.Body, e.Comment.HtmlUrl, commentAuthor, response))
+	}
+
+	if err := gc.ReopenIssue(org, repo, number); err != nil {
+		return err
+	}
+	// Add a comment after reopening the issue to leave an audit trail of who
+	// asked to reopen it.
+	return gc.CreateGiteeIssueComment(
+		org, repo, number, plugins.FormatResponseRaw(e.Comment.Body, e.Comment.HtmlUrl, commentAuthor, "Reopened this issue."))
```
Comment (zengchen1024) [inline prow/gitee-plugins/lifecycle/reopen.go:30]: response := plugins.FormatResponseRaw(e.Comment.Body, e.Comment.HtmlUrl, commentAuthor, "Closed this PR.")
	return gc.CreatePRComment(org, repo, number, response)

File: prow/gitee-plugins/lifecycle/close.go [hunk 26]
```diff
@@ -0,0 +1,67 @@
...
+	if ne.CommenterIsAuthor() {
+		return true
+	}
+	commenter := ne.GetCommenter()
+	isCollaborator, err := gh.IsCollaborator(org, repo, commenter)
+	if err != nil {
+		log.WithError(err).Errorf("Failed IsCollaborator(%s, %s, %s)", org, repo, commenter)
+	}
+	return isCollaborator
+}
```
Comment (zengchen1024) [inline prow/gitee-plugins/lifecycle/close.go:67]: func isAuthorOrCollaborator(org, repo, author string, ne *gitee.NoteEvent, gh giteeClient, log *logrus.Entry) bool {
	if author == ne.Commenter() {
		return true
	}
	isCollaborator, err := gh.IsCollaborator(org, repo, ne.Commenter())
	if err != nil {
		log.WithError(err).Errorf("Failed IsCollaborator(%s, %s, %s)", org, repo,  ne.Commenter())
	}
	return isCollaborator
}

File: prow/gitee/BUILD.bazel [hunk 27]
```diff
@@ -8,6 +8,7 @@ go_library(
         "github.go",
         "interface.go",
         "webhooks.go",
+        "noteevent.go",
```
Comment (zengchen1024) [inline prow/gitee/BUILD.bazel:11]: can change another name

File: prow/gitee-plugins/lifecycle/lifecycle.go [hunk 28]
```diff
@@ -0,0 +1,109 @@
...
+	}
+
+	if ne.IsIssue() {
+		return l.handleIssue(e, log)
+	}
+
+	return nil
+}
+
+func (l *lifecycle) handleIssue(e *sdk.NoteEvent, log *logrus.Entry) error {
```
Comment (zengchen1024) [inline prow/gitee-plugins/lifecycle/lifecycle.go:99]: s/e/ne

**Pull request merged**